### ISSUE TITLE
ng_di and ECMAScript 2015

### ISSUE THREAD

**Issue by britter** [Gillespie59/eslint-plugin-angular#138]
I'm using babel.js to transpile ECMAScript 2015 code to JavaScript. My ECMAScript looks something like this:

```javascript
class MyController {

    constructor(DataStore) {
        this.dataStore = DataStore;
    }

    toggle() {
        this.active = !this.active;
        this.dataStore.addData('active', this.active);
    }
}

angular.module('myModule')
  .controller('MyController', ['DataStore', MyController]);
```

bable.js will transpile this to something like:

```javascript
var MyController = (function () {
    function MyController(DataStore) {
        _classCallCheck(this, MyController);

        this.dataStore = DataStore;
    }

    _createClass(MyController, [{
        key: 'toggle',
        value: function toggle() {
            this.active = !this.active;
            this.dataStore.addData('active', this.active);
        }
    }]);

    return MyController;
})();

angular.module('myModule')
   .controller('BausteinController', ['DataStore', BausteinController]);
```

The transpiled code looks correct to me, nevertheless ng_di will be reported. Am I missing something? Is there a way to make eslint-plugin-angular aware of ECMAScript 2015?

**Comment by bripkens** [comment]
The default configuration for `ng_di` is `'ng_di': [2, 'function']`. It looks like you are using the array notation though. Could you please try again with `'ng_di': [2, 'array']`?

**Comment by bripkens** [comment]
Also, when using ES2015, you can use a [shorter syntax](https://blog.codecentric.de/en/2015/03/five-ways-to-write-better-angular-services/). For instance:

```javascript
class MyController {

    static get $inject() {
        return ['DataStore'];
    }

    constructor(DataStore) {
        this.dataStore = DataStore;
    }

    toggle() {
        this.active = !this.active;
        this.dataStore.addData('active', this.active);
    }
}
```

This way both dependency specifications are located near each other and updates are easier. Unfortunately it does not look like the `eslint-plugin-angular` does [not support](https://github.com/Gillespie59/eslint-plugin-angular/blob/d751e5538d39a60677e5ed2a4c540487789064bb/rules/ng_di.js#L16-L28) the property based dependency specification.

**Comment by Gillespie59** [comment]
Yes we do not expect this syntax for the moment. 
We should be able to detect if we are using ES5 or ES2015 syntax (constructor functions or classes)

If classes are used, do the same check implemented in the ng_di rule, but with the getter of the $inject property.

**Comment by bripkens** [comment]
Actually you do not need to detect ES2015 as this is standard Angular behavior supported in all JavaScript versions. See the documentation of `$injector` for information about the [`$inject` property](https://docs.angularjs.org/api/auto/service/$injector).

Meaning: The following code should be sufficient for the `ng_di` rule:

```
if(syntax === 'property' && (!utils.isFunctionType(param) || !paramHas$injectProperty) {
  report...
}
```

**Comment by Gillespie59** [comment]
You are totally right @bripkens . We can easily add the $inject property in our rule. 


But, a more general question, should we start supporting ES2015 syntax ? like class ... I think so . But maybe we can create a second issue for that.

**Comment by bripkens** [comment]
@Gillespie59 ES2015 support sounds great, especially since eslint supports ES2015.

**Comment by britter** [comment]
> The default configuration for ng_di is 'ng_di': [2, 'function']. It looks like you are using the array notation though. Could you please try again with 'ng_di': [2, 'array']?

This fixes my problem. Although I'd like to be able to use the property dependency specification as suggested by @bripkens. It fits better into EC2015 code.

**Comment by britter** [comment]
Closing this, because the problem can be fixed by using `'ng_di': [2, 'array']` in .eslintrc

Thank you!
